File: frontend/mobile/lib/core/services/prayer_times_service.dart
import 'package:dio/dio.dart';
import '../network/dio_client.dart';
import '../network/api_endpoints.dart';

/// Model for Prayer Times
class PrayerTimes {
  final String fajr;
  final String sunrise;
  final String dhuhr;
  final String asr;
  final String maghrib;
  final String isha;
  final DateTime date;
  final String location;

  PrayerTimes({
    required this.fajr,
    required this.sunrise,
    required this.dhuhr,
    required this.asr,
    required this.maghrib,
    required this.isha,
    required this.date,
    required this.location,
  });

  factory PrayerTimes.fromJson(Map<String, dynamic> json) {
    return PrayerTimes(
      fajr: json['fajr'] ?? '',
      sunrise: json['sunrise'] ?? '',
      dhuhr: json['dhuhr'] ?? '',
      asr: json['asr'] ?? '',
      maghrib: json['maghrib'] ?? '',
      isha: json['isha'] ?? '',
      date: DateTime.parse(json['date'] ?? DateTime.now().toIso8601String()),
      location: json['location'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'fajr': fajr,
      'sunrise': sunrise,
      'dhuhr': dhuhr,
      'asr': asr,
      'maghrib': maghrib,
      'isha': isha,
      'date': date.toIso8601String(),
      'location': location,
    };
  }

  /// Get the next prayer name and time
  Map<String, String> getNextPrayer() {
    final now = DateTime.now();
    final prayers = [
      {'name': 'الفجر', 'time': fajr},
      {'name': 'الشروق', 'time': sunrise},
      {'name': 'الظهر', 'time': dhuhr},
      {'name': 'العصر', 'time': asr},
      {'name': 'المغرب', 'time': maghrib},
      {'name': 'العشاء', 'time': isha},
    ];

    for (var prayer in prayers) {
      final prayerTime = _parseTime(prayer['time']!);
      if (prayerTime.isAfter(now)) {
        return {'name': prayer['name']!, 'time': prayer['time']!};
      }
    }

    // If all prayers passed, return Fajr of next day
    return {'name': 'الفجر', 'time': fajr};
  }

  DateTime _parseTime(String time) {
    final parts = time.split(':');
    final now = DateTime.now();
    return DateTime(
      now.year,
      now.month,
      now.day,
      int.parse(parts[0]),
      int.parse(parts[1]),
    );
  }
}

/// Model for Hijri Date
class HijriDate {
  final int day;
  final int month;
  final int year;
  final String monthName;
  final String weekday;

  HijriDate({
    required this.day,
    required this.month,
    required this.year,
    required this.monthName,
    required this.weekday,
  });

  factory HijriDate.fromJson(Map<String, dynamic> json) {
    return HijriDate(
      day: json['day'] ?? 1,
      month: json['month'] ?? 1,
      year: json['year'] ?? 1445,
      monthName: json['month_name'] ?? '',
      weekday: json['weekday'] ?? '',
    );
  }

  String get formatted => '$weekday، $day $monthName $year هـ';
}

/// Prayer Times Service
class PrayerTimesService {
  final DioClient _dioClient;

  PrayerTimesService(this._dioClient);

  /// Get prayer times for a specific location
  Future<PrayerTimes> getPrayerTimes({
    required double latitude,
    required double longitude,
    String? madhab,
  }) async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.prayerTimes,
        queryParameters: {
          'latitude': latitude,
          'longitude': longitude,
          if (madhab != null) 'madhab': madhab,
        },
      );

      return PrayerTimes.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to fetch prayer times: $e');
    }
  }

  /// Get Hijri date for today
  Future<HijriDate> getHijriDate() async {
    try {
      final response = await _dioClient.get(ApiEndpoints.hijriDate);
      return HijriDate.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to fetch Hijri date: $e');
    }
  }

  /// Get monthly prayer times
  Future<List<PrayerTimes>> getMonthlyPrayerTimes({
    required double latitude,
    required double longitude,
    required int month,
    required int year,
    String? madhab,
  }) async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.monthlyPrayerTimes,
        queryParameters: {
          'latitude': latitude,
          'longitude': longitude,
          'month': month,
          'year': year,
          if (madhab != null) 'madhab': madhab,
        },
      );

      return (response.data as List)
          .map((json) => PrayerTimes.fromJson(json))
          .toList();
    } catch (e) {
      throw Exception('Failed to fetch monthly prayer times: $e');
    }
  }
}

File: frontend/nextjs-app/src/lib/services/prayer-times-service.ts
import { axiosClient } from '../api/axios-client';
import { API_ENDPOINTS } from '../api/endpoints';

export interface PrayerTimes {
  fajr: string;
  sunrise: string;
  dhuhr: string;
  asr: string;
  maghrib: string;
  isha: string;
  date: string;
  location: string;
}

export interface HijriDate {
  day: number;
  month: number;
  year: number;
  monthName: string;
  weekday: string;
}

export interface NextPrayer {
  name: string;
  time: string;
  timeRemaining: {
    hours: number;
    minutes: number;
    seconds: number;
  };
}

export class PrayerTimesService {
  /**
   * Get prayer times for a specific location
   */
  static async getPrayerTimes(
    latitude: number,
    longitude: number,
    madhab?: string
  ): Promise<PrayerTimes> {
    const response = await axiosClient.get<PrayerTimes>(
      API_ENDPOINTS.PRAYER_TIMES,
      {
        params: {
          latitude,
          longitude,
          ...(madhab && { madhab }),
        },
      }
    );
    return response.data;
  }

  /**
   * Get Hijri date for today
   */
  static async getHijriDate(): Promise<HijriDate> {
    const response = await axiosClient.get<HijriDate>(
      API_ENDPOINTS.HIJRI_DATE
    );
    return response.data;
  }

  /**
   * Get monthly prayer times
   */
  static async getMonthlyPrayerTimes(
    latitude: number,
    longitude: number,
    month: number,
    year: number,
    madhab?: string
  ): Promise<PrayerTimes[]> {
    const response = await axiosClient.get<PrayerTimes[]>(
      API_ENDPOINTS.MONTHLY_PRAYER_TIMES,
      {
        params: {
          latitude,
          longitude,
          month,
          year,
          ...(madhab && { madhab }),
        },
      }
    );
    return response.data;
  }

  /**
   * Calculate next prayer and time remaining
   */
  static getNextPrayer(prayerTimes: PrayerTimes): NextPrayer {
    const now = new Date();
    const prayers = [
      { name: 'الفجر', time: prayerTimes.fajr },
      { name: 'الشروق', time: prayerTimes.sunrise },
      { name: 'الظهر', time: prayerTimes.dhuhr },
      { name: 'العصر', time: prayerTimes.asr },
      { name: 'المغرب', time: prayerTimes.maghrib },
      { name: 'العشاء', time: prayerTimes.isha },
    ];

    for (const prayer of prayers) {
      const prayerTime = this.parseTime(prayer.time);
      if (prayerTime > now) {
        const diff = prayerTime.getTime() - now.getTime();
        return {
          name: prayer.name,
          time: prayer.time,
          timeRemaining: {
            hours: Math.floor(diff / (1000 * 60 * 60)),
            minutes: Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60)),
            seconds: Math.floor((diff % (1000 * 60)) / 1000),
          },
        };
      }
    }

    // If all prayers passed, return Fajr of next day
    const fajrTime = this.parseTime(prayerTimes.fajr);
    fajrTime.setDate(fajrTime.getDate() + 1);
    const diff = fajrTime.getTime() - now.getTime();

    return {
      name: 'الفجر',
      time: prayerTimes.fajr,
      timeRemaining: {
        hours: Math.floor(diff / (1000 * 60 * 60)),
        minutes: Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60)),
        seconds: Math.floor((diff % (1000 * 60)) / 1000),
      },
    };
  }

  /**
   * Parse time string to Date object
   */
  private static parseTime(time: string): Date {
    const [hours, minutes] = time.split(':').map(Number);
    const date = new Date();
    date.setHours(hours, minutes, 0, 0);
    return date;
  }

  /**
   * Format Hijri date
   */
  static formatHijriDate(hijriDate: HijriDate): string {
    return `${hijriDate.weekday}، ${hijriDate.day} ${hijriDate.monthName} ${hijriDate.year} هـ`;
  }
}
